Add Task, which will be basis for persistent state.

gfr10598 opened the pull request:
This introduces "Task", which stores info about the state of a single gardener reprocess/dedup task.  It also changes the channels between GoRoutines to handle Tasks instead of strings.

Future PRs will add more functionality.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/etl-gardener/46)
<!-- Reviewable:end -->

stephen-soltesz commented:
Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---

*[cloud/tq/queuehandler_test.go, line 24 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYF_gp6ERTf2sXKMg-:-LAYF_gp6ERTf2sXKMg0:b-je5ll6) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/cloud/tq/queuehandler_test.go#L24)):*
> ```Go
> 	}
> 	for i := 19; i < 29; i++ {
> 		cqh.Sink() <- state.Task{Name: fmt.Sprintf("gs://archive-mlab-test/ndt/2017/09/%2d/", i)}
> ```

The convention we're using throughout all projects is to suffix buckets with the project name containing them. Please rename this bucket to "archive-mlab-testing".

---

*[dispatch/deduphandler.go, line 169 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYHVa76ney5Gkd-ZHk:-LAYHVa76ney5Gkd-ZHl:b-84mitz) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/dispatch/deduphandler.go#L169)):*
> ```Go
> 				metrics.FailCount.WithLabelValues("NewDataset")
> 				log.Println(err)
> 				continue
> ```

Does this potentially leak valid tasks? Does that matter?

---

*[state/state.go, line 15 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYJQkoE7EtIBq7IF_U:-LAYJQkoE7EtIBq7IF_V:b-sjjxem) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L15)):*
> ```Go
> const (
> 	Initializing  State = iota
> 	Queuing             // Queuing all tasks for the task.
> ```

This sentence is confusing. I'm guessing the task types are different? e.g. a `TaskQueue` task vs a `state.Task`?

---

*[state/state.go, line 29 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYKNlrCRe0yXkZYiW9:-LAYKNlrCRe0yXkZYiWA:b-1bawee) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L29)):*
> ```Go
> }
> 
> // DatastoreSaver will implement a Saver that stores Task state in DataStore.
> ```

nit: the product name is "Datastore" without a capital "S".

---

*[state/state.go, line 44 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYKoyN97nPlSWUpAul:-LAYKoyN97nPlSWUpAum:b-lc19ux) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L44)):*
> ```Go
> type Task struct {
> 	Name    string // e.g. gs://archive-mlab-oti/ndt/2017/06/01/
> 	Suffix  string // e.g. 20170601
> ```

table suffix?

---

*[state/state.go, line 46 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYKt8i1lHGlWl60WT4:-LAYKt8i1lHGlWl60WT5:bpzp3qr) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L46)):*
> ```Go
> 	Suffix  string // e.g. 20170601
> 	State   State
> 	Queue   string // The queue the tasks were submitted to, or empty.
> ```

The string "empty" or the empty string?

---

*[state/state.go, line 47 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYL14RBtVHUvcrekTW:-LAYL14RBtVHUvcrekTX:b7ngud0) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L47)):*
> ```Go
> 	State   State
> 	Queue   string // The queue the tasks were submitted to, or empty.
> 	JobID   string // JobID, when the state is Deduplicating
> ```

nit: BigQuery JobID.

---

*[state/state.go, line 48 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYLCtWEZtPLrJoDwKG:-LAYLCtWEZtPLrJoDwKH:b-iajwep) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L48)):*
> ```Go
> 	Queue   string // The queue the tasks were submitted to, or empty.
> 	JobID   string // JobID, when the state is Deduplicating
> 	Err     error  // Processing error, if any
> ```

"Processing" in the general sense or in the `State` constant sense? Is in the Processing state the only time an error can occur?

---

*[state/state_test.go, line 11 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAZ1IMp8Nt7H2K-Pbts:-LAZ1IMp8Nt7H2K-Pbtt:b1kkpd0) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state_test.go#L11)):*
> ```Go
> )
> 
> type saver struct {
> ```

Please name this type `fakeSaver` or `testSaver`, or similar.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions.

---

*[cloud/tq/queuehandler_test.go, line 24 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYF_gp6ERTf2sXKMg-:-LAZwv2P1wc3CF3wO5jb:b-5inskd) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/cloud/tq/queuehandler_test.go#L24)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

The convention we're using throughout all projects is to suffix buckets with the project name containing them. Please rename this bucket to "archive-mlab-testing".
</blockquote></details>

Done.  Also fixed in other tq tests.

---

*[dispatch/deduphandler.go, line 169 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYHVa76ney5Gkd-ZHk:-LAZx6voEwEORGaonawm:b-s2y5v7) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/dispatch/deduphandler.go#L169)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Does this potentially leak valid tasks? Does that matter?
</blockquote></details>

Yes it does, but not concerned about this for now.  If we see these errors, we can pursue.  If there are more than a few, the sanity check will prevent the date partition from being committed to public tables.
Will add TODO though.

---

*[state/state.go, line 15 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYJQkoE7EtIBq7IF_U:-LAZxpVE5lEVM-m-c3GD:bj4un6m) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L15)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

This sentence is confusing. I'm guessing the task types are different? e.g. a `TaskQueue` task vs a `state.Task`?
</blockquote></details>

Improved comment.

---

*[state/state.go, line 29 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYKNlrCRe0yXkZYiW9:-LAZxujs49c0MePpf2Ys:b-896fix) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L29)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

nit: the product name is "Datastore" without a capital "S".
</blockquote></details>

Done.

---

*[state/state.go, line 44 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYKoyN97nPlSWUpAul:-LAZxxtO4Zie3hSXwf3q:b-896fix) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L44)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

table suffix?
</blockquote></details>

Done.

---

*[state/state.go, line 46 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYKt8i1lHGlWl60WT4:-LAZy8RiDHMYKvxpv3Qb:b4vgvyz) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L46)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

The string "empty" or the empty string?
</blockquote></details>

Replaced with ""

---

*[state/state.go, line 47 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYL14RBtVHUvcrekTW:-LAZyAUEEB7G8RKhE4E-:b-896fix) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L47)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

nit: BigQuery JobID.
</blockquote></details>

Done.

---

*[state/state.go, line 48 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYLCtWEZtPLrJoDwKG:-LAZyF2KD5OIdzcgRqfb:b-ie7tbk) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L48)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

"Processing" in the general sense or in the `State` constant sense? Is in the Processing state the only time an error can occur?
</blockquote></details>

Changed to "task handling"

---

*[state/state_test.go, line 11 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAZ1IMp8Nt7H2K-Pbts:-LAZylHg7EfYIlOfF0gj:b-896fix) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state_test.go#L11)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Please name this type `fakeSaver` or `testSaver`, or similar.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
PTAL.

---

Review status: 0 of 10 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-:-LAZz1W0BzOfxnf3SPvR:b-83330m)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[state/state.go, line 48 at r1](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LAYLCtWEZtPLrJoDwKG:-LA_9tN6618eMaotDFZn:b-u5gbiz) ([raw file](https://github.com/m-lab/etl-gardener/blob/66eeffb099ca12a6ce5492bb4c8764d42fc2786c/state/state.go#L48)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Comment text on Err looks unchanged.
</blockquote></details>

Hmm.  fixed.

---

*[state/state.go, line 14 at r2](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46#-LA_8OCB1nYXlaaTMIFU:-LA_AKy69R_WdEBQ9iTE:b-896fix) ([raw file](https://github.com/m-lab/etl-gardener/blob/bede0d2ed50dbb285a567c8a5a119fdc22eb7ac5/state/state.go#L14)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Just thinking out loud; what do you think of making the initial iota value be an invalid state? This way a zero-value Task would be invalid rather than "Initializing".
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/m-lab/etl-gardener/46)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Part of Story #37
